cirq/contrib/acquaintance/executor.py

```diff
@@ -0,0 +1,126 @@
+        self.mapping = execution_strategy.initial_mapping.copy()
+
+    def __call__(self, strategy: Circuit):
+        if not isinstance(strategy, AcquaintanceStrategy):
+            raise ValueError('Can only execute an acquaintance strategy.')
+        expose_acquaintance_gates(strategy)
+        super().optimize_circuit(strategy)
+        return self.mapping.copy()
+
+    def optimization_at(self, circuit: Circuit, index: int, op: Operation):
+        if (isinstance(op, GateOperation) and
+                isinstance(op.gate, AcquaintanceOpportunityGate)):
+            logical_indices = tuple(self.mapping[q] for q in op.qubits)
+            logical_operations = self.execution_strategy.get_operations(
+                    logical_indices, op.qubits)
+            clear_span = int(not self.execution_strategy.keep_acquaintance)
+
+            return PointOptimizationSummary(
+                    clear_span=clear_span,
+                    clear_qubits=op.qubits,
+                    new_operations=logical_operations)
+
+        if (isinstance(op, GateOperation) and
+                isinstance(op.gate, PermutationGate)):
+            op.gate.update_mapping(self.mapping, op.qubits)
+            return
+
+        raise TypeError('Can only execute a strategy consisting of gates that '
+                         'are instances of AcquaintanceOpportunityGate or '
+                         'PermutationGate.')
+
+class GreedyExecutionStrategy(ExecutionStrategy):
+    def __init__(self,
+                 gates: LogicalGates,
+                 initial_mapping: LogicalMapping
+                 ) -> None:
+        if len(set(len(indices) for indices in gates)) > 1:
+            raise NotImplementedError(
+                    'Can only implement greedy strategy if all gates '
+                    'are of the same arity.')
+        self.index_set_to_gates = self.canonicalize_gates(gates)
+        self._initial_mapping = initial_mapping.copy()
+
+    @property
+    def initial_mapping(self) -> LogicalMapping:
+        return self._initial_mapping
+
+    def get_operations(self,
+                       indices: Sequence[LogicalIndex],
+                       qubits: Sequence[QubitId]
+                       ) -> OP_TREE:
+        index_set = frozenset(indices)
+        if index_set in self.index_set_to_gates:
+            gates = self.index_set_to_gates.pop(index_set)
+            index_to_qubit = dict(zip(indices, qubits))
+            for gate_indices, gate in gates.items():
+                yield gate(*[index_to_qubit[i] for i in gate_indices])
+        return
+
+
+    @staticmethod
+    def canonicalize_gates(gates: LogicalGates
+        ) -> Dict[frozenset, LogicalGates]:
+        canonicalized_gates = defaultdict(dict
+            ) # type: DefaultDict[frozenset, LogicalGates]
+        for indices, gate in gates.items():
+            indices = tuple(indices)
+            canonicalized_gates[frozenset(indices)][indices] = gate
```

Comment:
It's intended for this to overwrite old values when the operations have the same index set?

Reply:
I'm not sure what you mean. `indices` is ordered, and unique given that they are keys of a `dict`.

The idea is to go from "unordered index set" to "ordered index tuple" to gate.